### PR TITLE
Devel/cphi gate

### PR DESCRIPTION
**Pull request by nathlacroix**
Changes proposed in this pull request:
- Repeat patterns handled more smoothly. Note: will break XLD repeat patterns until they update their sequence to the new format
- Arbitrary Sequence measurement
- prototype Arbitrary phase gate
- Reloading of arrays list and dict parameters
- analysis v3, probably not the latest version

@antsr

### CODE REVIEW

**Review by stephlazar — CHANGES_REQUESTED** [review]
Can you please remove the print statements in:

- MC at lines 1271, 1272
- ro_analysis at line 1465

File: pycqed/analysis_v2/readout_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/readout_analysis.py:1414]: Why did the key change here?

The key `'measured_values_ord_dict'` is still used in 4 files, most of which is DCL code that we don't use as far as I know. Can we get rid of `'measured_values_ord_dict'` everywhere now?

File: pycqed/analysis_v2/readout_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/readout_analysis.py]: Delete `print` statement. Can replace with `log.debug` statement

File: pycqed/analysis_v2/spectroscopy_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/spectroscopy_analysis.py]: Remove `print` or replace with `log.debug`

File: pycqed/analysis_v2/spectroscopy_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/spectroscopy_analysis.py]: `logging` -> `log`

File: pycqed/analysis_v2/spectroscopy_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/spectroscopy_analysis.py]: `logging` -> `log`

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/timedomain_analysis.py:210]: why not `self.metadata = self.raw_data_dict.get('exp_metadata', {})`?

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [inline pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py]: Basis in what space, this is left a bit confusing to me.

File: pycqed/measurement/detector_functions.py
Comment (antsr) [inline pycqed/measurement/detector_functions.py]: `print` -> `log.debug`

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [inline pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:2499]: `print` -> `log.error`

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [inline pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:2497]: Why do we run the base analysis here? Don't we have an analysis for normal RB?

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [inline pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:1323]: Why do we run the base analysis here? Don't we have an analysis for normal RB?

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [inline pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py]: `print` -> `log.error`

File: pycqed/analysis_v3/saving.py
Comment (antsr) [inline pycqed/analysis_v3/saving.py]: `print` -> `log.info`

File: pycqed/analysis_v3/saving.py
Comment (antsr) [inline pycqed/analysis_v3/saving.py]: `print` -> `log.info`

File: pycqed/analysis_v3/saving.py
Comment (antsr) [inline pycqed/analysis_v3/saving.py]: `print` -> `log.info`

File: pycqed/analysis_v3/saving.py
Comment (antsr) [inline pycqed/analysis_v3/saving.py]: `print` -> `log.info`

File: pycqed/analysis_v3/saving.py
Comment (antsr) [inline pycqed/analysis_v3/saving.py]: `print` -> `log.info`

File: pycqed/analysis_v2/readout_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/readout_analysis.py]: remove `print` statement

File: pycqed/analysis_v2/readout_analysis.py
Comment (antsr) [inline pycqed/analysis_v2/readout_analysis.py]: `print` -> `log.debug`

File: pycqed/analysis/measurement_analysis.py
Comment (antsr) [inline pycqed/analysis/measurement_analysis.py]: `print` -> `log.debug`

File: pycqed/measurement/detector_functions.py
Comment (antsr) [inline pycqed/measurement/detector_functions.py]: `print` -> `log.info`

File: pycqed/measurement/measurement_control.py
Comment (antsr) [inline pycqed/measurement/measurement_control.py:1135]: Does this avoid `'[1,2,...,100]'` strings?

File: pycqed/measurement/measurement_control.py
Comment (antsr) [inline pycqed/measurement/measurement_control.py]: `log.debug`?

File: pycqed/measurement/multi_qubit_module.py
Comment (antsr) [inline pycqed/measurement/multi_qubit_module.py:248]: Why was this removed?

File: pycqed/measurement/multi_qubit_module.py
Comment (antsr) [inline pycqed/measurement/multi_qubit_module.py:427]: Since dict is a mutable object, the default value can be changed during runtime. Better to set `None` as the default and check for `is None`.

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:926]: This is not really `qudev_master` level code, yes :D
I would guess we can still leave it in for now.

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py]: `print` -> `log.debug`

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:878]: @nathlacroix Maybe you could update this function to use the new repeat pattern syntax. This should be the only place where we use it on XLD.

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py]: `print` -> `log.debug`

File: pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:56]: Nice and compact code here. Does it work?

File: pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:122]: `t1_active_reset` -> `t1_seq`?

File: pycqed/measurement/waveform_control/pulsar.py
Comment (antsr) [inline pycqed/measurement/waveform_control/pulsar.py]: remove

File: pycqed/measurement/waveform_control/segment.py
Comment (antsr) [inline pycqed/measurement/waveform_control/segment.py]: `print` -> `log.error`

File: pycqed/measurement/waveform_control/segment.py
Comment (antsr) [inline pycqed/measurement/waveform_control/segment.py]: `print` -> `log.error`

File: pycqed/utilities/general.py
Comment (antsr) [inline pycqed/utilities/general.py:238]: 👍

File: pycqed/utilities/general.py
Comment (antsr) [inline pycqed/utilities/general.py]: `print` -> `log.error`

File: pycqed/measurement/waveform_control/pulsar.py
Comment (antsr) [inline pycqed/measurement/waveform_control/pulsar.py]: `print` -> `log.error`

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (antsr) [inline pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py]: remove

File: pycqed/measurement/multi_qubit_module.py
Comment (antsr) [inline pycqed/measurement/multi_qubit_module.py]: `print` -> `log.debug`

File: pycqed/measurement/multi_qubit_module.py
Comment (antsr) [inline pycqed/measurement/multi_qubit_module.py]: `print` -> `log.debug`

**Review by antsr — CHANGES_REQUESTED** [review]
@nathlacroix Maybe you can go through the comments and fix them as you see fit and reference @Vintrix where you think that she should make the changes

File: pycqed/measurement/detector_functions.py
Comment (stephlazar) [reply pycqed/measurement/detector_functions.py]: This statement is removed in Proj_XLD.

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (stephlazar) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:2499]: Additional support has been added in Proj_XLD.

File: pycqed/analysis_v2/readout_analysis.py
Comment (nathlacroix) [reply pycqed/analysis_v2/readout_analysis.py:1414]: @Vintrix  do you know ? I didnt change this myself.

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (stephlazar) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:2497]: This has been fixed in Proj_XLD. The code here is a preliminary version. 
BaseDataAnalysis here is now called PipelineDataAnalysis; it is the one that handles the pipe in analysis_v3, so this is how one runs any analysis in analysis_v3. The only part specific to each analysis is the pipeline.

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (stephlazar) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:1323]: See above comment.

File: pycqed/measurement/measurement_control.py
Comment (stephlazar) [reply pycqed/measurement/measurement_control.py]: No, remove! I used this for debugging.

File: pycqed/measurement/multi_qubit_module.py
Comment (stephlazar) [reply pycqed/measurement/multi_qubit_module.py:248]: Because it's not needed. The classifier detector has 4 different options currently on Proj_XLD: averaged, correlated, thresholded, classified. These options are specified by the user in "get_values_function_kwargs." This seemed cleaner than to make 4 different instances of that detector function, but is probably not be the best solution.

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (nathlacroix) [reply pycqed/analysis_v2/timedomain_analysis.py:210]: Change made by @Vintrix . Was because if exp_metadata is empty / there is no meta data, an empty list is returned, therefore it does not set it to the wanted value of {}.
I agree that a better fix would be to check why an empty list is returned by the load hdf

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (nathlacroix) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py]: eg. ['ge', 'gf'] or ['ge', 'ortho'] describing how weights were computed from the traces.
'ge' indicates (trace['g'] - trace['e']) and ortho refers to orthogonal to the previous (and spanning the 3 states)

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (nathlacroix) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:2499]: will not change then

File: pycqed/measurement/measurement_control.py
Comment (nathlacroix) [reply pycqed/measurement/measurement_control.py:1135]: yes

File: pycqed/measurement/multi_qubit_module.py
Comment (nathlacroix) [reply pycqed/measurement/multi_qubit_module.py:427]: We said it would be ok for now since unpacked

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (nathlacroix) [reply pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:878]: @antr not sure I fully understand which channels which pulses should be repeated here, since it is a bit hacky ;) I would rather have Christian do it. The doc string of the sequence.repeat() is quite straight forward.

File: pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py
Comment (nathlacroix) [reply pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:56]: thats a wrapper of sequence.repeat() assuming all readouts are the same and thus we can use the readout pattern (n_acq_elements, 1). We should discuss further with christian but I think this should be ok for all single qubits experiments. Might be more difficult for multiqubit experiments, there I am not sure how it should be done because I don't have a perfect understanding of how the repeat pattern works.

File: pycqed/measurement/waveform_control/pulsar.py
Comment (nathlacroix) [reply pycqed/measurement/waveform_control/pulsar.py]: No, this is important to be able to plot segments without connection to the setup. Needs a dummy uhf and this must be an accepted type in PulsarUHF

File: pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py
Comment (nathlacroix) [reply pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:122]: This naming is used to differentiate sequence which are in the new framework and handle both ge and ef pulses. Once we have removed all the old ones (which we should, I already did most of them but Im not sure in the remaining ones which ones could still be used), then I will rename all of the sequences :)

File: pycqed/analysis_v2/readout_analysis.py
Comment (antsr) [reply pycqed/analysis_v2/readout_analysis.py:1414]: @Vintrix can we get rid of `measured_values_ord_dict` everywhere? Will you do it or should I do it?

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py]: ^ @nathlacroix add this example to the docstring maybe, since the expected input is also not clear from the validator.

File: pycqed/analysis_v2/timedomain_analysis.py
Comment (antsr) [reply pycqed/analysis_v2/timedomain_analysis.py:210]: I trust this works then. Its not nice, but I think we can leave it for now.

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py:2497]: I see, and it gets the pipeline from the metadata of the hdf5.

File: pycqed/measurement/measurement_control.py
Comment (antsr) [reply pycqed/measurement/measurement_control.py:1135]: :+1:

File: pycqed/measurement/multi_qubit_module.py
Comment (antsr) [reply pycqed/measurement/multi_qubit_module.py:248]: Ok, thats good, I was just curious.

File: pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py
Comment (antsr) [reply pycqed/measurement/pulse_sequences/multi_qubit_tek_seq_elts.py:878]: ok, we need to make Christian make a github account then 😆

File: pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py
Comment (antsr) [reply pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:122]: Ok, agreed that the different naming convention is a good clue for the framework version it uses.

Comment (antsr) [reply pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:122]: I guess we could `class T1(swf.Hard_Sweep)` then, since it references a nonexistant `sqs.T1_seq`. Also other AWG sweep functions that reference sequence functions that don't exist.

File: pycqed/measurement/waveform_control/pulsar.py
Comment (antsr) [reply pycqed/measurement/waveform_control/pulsar.py]: Sorry for the laconic comment before. 
This is a duplicate import as the dummyUHFQC is imported on line 45-46

Comment (antsr) [reply pycqed/measurement/waveform_control/pulsar.py]: I will fix

File: pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py
Comment (antsr) [reply pycqed/measurement/pulse_sequences/single_qubit_tek_seq_elts.py:122]: I will remove the `class T1`

File: pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py
Comment (antsr) [reply pycqed/instrument_drivers/meta_instrument/qubit_objects/QuDev_transmon.py]: I will fix myself.

File: pycqed/analysis_v2/readout_analysis.py
Comment (stephlazar) [reply pycqed/analysis_v2/readout_analysis.py:1414]: Fixed.

**Review by antsr — APPROVED** [review]

**Comment by antsr** [comment]
You should approve changes if you approve them @Vintrix

**Review by stephlazar — APPROVED** [review]

**Comment by antsr** [comment]
🥇 🎉

**Pull request merged**
